### PR TITLE
Watch for report file changes

### PR DESCRIPTION
**Pull request by nderwin**
* look in `src/main/resources` for any `.jrxml` files, and add them to the watched paths for hot deployment

Signed-off-by:Nathan Erwin <nathan.d.erwin@gmail.com>

### CODE REVIEW

**Comment by nderwin** [comment]
Fixes #70

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.Iterator;
+import java.util.Set;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+/**
+ * The set of report root paths.
+ */
+public final class ReportRootsBuildItem extends SimpleBuildItem implements Iterable<String> {
```
Comment (gastaldi) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java:11]: This class doesn't seem to be useful, you can consume `List<ReportRootBuildItem>` instead?

**Review by gastaldi — CHANGES_REQUESTED** [review]
Please add a test

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java (hunk 1 above)
Comment (nderwin) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java:11]: I was following the code from the Qute extension, which has it set up this way.  I can give it a go with simplifying it though.

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java [hunk 2]
```diff
@@ -14,12 +19,42 @@ class JasperReportsProcessor {
...
     }
 
+    @BuildStep
+    ReportRootBuildItem defaultReportRoot() {
+        return new ReportRootBuildItem(DEFAULT_ROOT_PATH);
+    }
+
+    @BuildStep
+    ReportRootsBuildItem collectReportRoots(List<ReportRootBuildItem> reportRoots) {
+        Set<String> roots = new HashSet<>();
```
Comment (melloware) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:36]: Maybe use LinkedHashSet to keep them in order?

**Review by melloware — CHANGES_REQUESTED** [review]

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java (hunk 2 above)
Comment (gastaldi) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:36]: Or maybe consume the List directly?

**Comment by nderwin** [comment]
> Please add a test

I might need some help on that...  I've been testing it by adding log messages, then running the integration project to see what I expect in the log...  but I'm not sure how to test the hot reload in the deployment project since it doesn't have any HTTP endpoints to call to force a reload.

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java (hunk 2 above)
Comment (nderwin) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:36]: Consuming the list directly, though we're only producing one item in that list... :thinking:

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java (hunk 1 above)
Comment (nderwin) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportRootsBuildItem.java:11]: Will be fixed in the next push.

**Comment by melloware** [comment]
@nderwin the Integration test has REST endpoints exposed?  Run `mvn quarkus:dev` and hit it from SwaggerUI.

image

**Comment by nderwin** [comment]
Yeah...  for some reason, I was thinking I needed the test in the deployment project, makes more sense in the integration-**_test_** project.

**Comment by nderwin** [comment]
Oh, so `Cannot mix both io.quarkus.test.QuarkusDevModeTest based tests and io.quarkus.test.junit.QuarkusTestExtension based tests in the same run` - I guess we can't test the hot reload in the `integration-tests` project then...

**Comment by melloware** [comment]
well the first thing i would do it try it manually

1. in integration test run `mvn quarkus:dev`
2. Hit the REST endpoint in the browser
3. Change the JRXML
4. Hit the rest endpoint again does it hot reload?

**Comment by nderwin** [comment]
@melloware Manual testing - yes, I've done that (with log messages) to prove that it restarts when I hit the REST endpoint because the .jrxml file changed.

I (and probably @gastaldi) was hoping to add some automated tests though.  I'm ok with forgetting about that for now, but I'm not in charge here... :grin:

**Review by melloware — APPROVED** [review]

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java [hunk 3]
```diff
@@ -23,12 +25,37 @@ class JasperReportsProcessor {
...
     FeatureBuildItem feature() {
         return new FeatureBuildItem(FEATURE);
     }
 
+    @BuildStep
+    ReportRootBuildItem defaultReportRoot() {
+        return new ReportRootBuildItem(DEFAULT_ROOT_PATH);
+    }
+
+    @BuildStep
```
Comment (melloware) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:40]:
```suggestion
    @BuildStep(onlyIf = IsDevelopment.class)
```

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java [hunk 4]
```diff
@@ -23,12 +25,37 @@ class JasperReportsProcessor {
...
     private static final String FEATURE = "jasperreports";
     private static final String EXTENSIONS_FILE = "jasperreports_extension.properties";
+    private static final String DEFAULT_ROOT_PATH = "";
 
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(FEATURE);
     }
 
+    @BuildStep
```
Comment (melloware) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:35]:
```suggestion
    @BuildStep(onlyIf = IsDevelopment.class)
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java (hunk 3 above)
Comment (melloware) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:40]: You will have to add the import too.

**Comment by melloware** [comment]
I will take care of these @nderwin

**Pull request merged**